### PR TITLE
Add query param abstraction

### PR DESCRIPTION
**Pull request by mullermp**
For query/rest xml, have a glorified hash wrapper builder thing.

### CODE REVIEW

File: hearth/lib/hearth/query/param_list.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      end
+
+      # @return [Array<Param>] Returns an array of sorted {Param} objects.
+      def to_a
+        @params.values.sort
+      end
+
+      # @return [String]
+      def to_s
+        to_a.map(&:to_s).join('&')
```
Comment (alextwoods) [inline hearth/lib/hearth/query/param_list.rb:53]: How does this handle list/set of params? the @httpQuery param requires us to do eg:
`foo = ["a", "b"]` serialize to `foo=a&foo=b`

https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#httpquery-trait

Comment (mullermp) [reply hearth/lib/hearth/query/param_list.rb:53]: Ah.. It doesn't. I mostly lifted from v3. Is that mainly rest?

Comment (alextwoods) [reply hearth/lib/hearth/query/param_list.rb:53]: Yeah, I believe only rest support the @httpQueryTrait.

File: hearth/lib/hearth/query/param.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+      # @return [Boolean]
+      def <=>(other)
+        name <=> other.name
+      end
+
+      private
+
+      def serialize(name, value)
+        value ? "#{escape(name)}=#{escape(value)}" : "#{escape(name)}="
```
[the remark this answers is not in the document]
Comment (mullermp) [reply hearth/lib/hearth/query/param.rb:45]: I decided to do them in the protocol builders. I think node/xml work similarly (expect strings)?

**Review by alextwoods — APPROVED** [review]

**Pull request merged**